### ISSUE TITLE
Address filtering is semi-broken

### ISSUE THREAD

**Issue by Stebalien** [libp2p/go-libp2p#841]
Due to https://github.com/multiformats/go-multiaddr-fmt/pull/1, the TCP, UDP, and WS(S) multiaddr matchers now match _DNS_ multiaddrs. This means we'll try to _dial_ DNS multiaddrs and succeed, even if we've blocked the underlying address/subnet address via the address filter.

In the TCP & WS transports, this isn't terrible because we check _again_ after establishing the connection. However, it looks like the QUIC transport uses the input address, not the final resolved address, so this second fallback check fails to work either.

TODO:

* [x] Expose _resolved_ addresses in QUIC.
* [ ] Fix the filtering logic, either in go-multiaddr-fmt, or in the individual transports.
* [ ] Add a test in libp2p.

**Comment by Stebalien** [comment]
Note: this is a blocker for the go-ipfs release.

**Comment by marten-seemann** [comment]
Not sure if I understand the issue correctly. Should this be an issue in go-libp2p-quic-transport?

Is it incorrect to use the `raddr` we dialed as a `RemoteMultiaddr` here? https://github.com/libp2p/go-libp2p-quic-transport/blob/c250617d5b6ef80e4641295d703bb7b514d34200/transport.go#L119-L172

**Comment by Stebalien** [comment]
I've filed an issue for the QUIC side of this https://github.com/libp2p/go-libp2p-quic-transport/issues/123

However, that's only showing up because we're passing dns addresses directly to transports. Historically, we we resolved them in the routed host. We should either go back to doing that (might cause issues with some recent changes to the websocket transport?) or resolve then filter addresses inside transports.

Tangentially related, we should filter inbound connections in QUIC before we even try to handshake: https://github.com/libp2p/go-libp2p-quic-transport/issues/124.

**Comment by raulk** [comment]
From an architectural perspective, the transport should receive the raw multiaddr and decide what to do with it.

Rationale: future transports like Ethernet, Bluetooth, Tor, etc. have their own address format, and they may support custom name resolution schemes that aren't DNS based.

Putting the resolution logic on the side of go-multiaddr leaks such concerns, and favours monolithic, god-like "know-it-all" logic.

I'd support the go-multiaddr family of repos offering functions to facilitate commonplace resolutions for transports to reuse (e.g. DNS resolution), but ultimately it is the transport that knows which addressing schemes are relevant to it, and is responsible for resolving what needs to be resolved.

**Comment by Stebalien** [comment]
@raulk To make sure we're on the same page, the bug here is that:

1. Currently, it's the routed host's responsibility to resolve DNS addresses.
2. Previously, the swarm would never actually try to _dial_ a DNS address.
  1. The swarm would call `transport.CanDial("/dns/...")`.
  2. The transport would call `mafmt.(TCP|UDP).Matches("/dns/...")` which would return _false_.
3. Now, `mafmt.(TCP|UDP).Matches("/dns/...") returns true.

This issue is _not_ about adding smarts to go-multiaddr.

> From an architectural perspective, the transport should receive the raw multiaddr and decide what to do with it.

For DNS, that might make sense. In general, that's not always possible:

* `/dnsaddr` addresses must be resolved before we can know the underlying transport.
* I wouldn't want every transport to have to know about every name system we decide to use.

> Rationale: future transports like Ethernet, Bluetooth, Tor, etc. have their own address format, and they may support custom name resolution schemes that aren't DNS based.

For transport-specific name systems, that totally makes sense.

> Putting the resolution logic on the side of go-multiaddr leaks such concerns, and favours monolithic, god-like "know-it-all" logic.

?

**Comment by Stebalien** [comment]
Digging into this, if we resolve inside the transports, we have several issues:

1. We may resolve to multiple addresses. Usually, it's up to the swarm to figure out how to schedule dials.
2. We'll end up dialing the same underlying addresses multiple times because, again, we do the scheduling/deduping inside the swarm itself.

Our options are to resolve in the swarm or resolve in the host (what we do now). Given that the swarm holds the transports, IMO, it would make more sense if the swarm _also_ held a set of pluggable resolvers.

**Comment by raulk** [comment]
I still regard address resolution as a transport concern, but not part of the transport dial. I’d enhance transports with an optional `ResolveAddr(Multiaddr) []Multiaddr` that the network can call to ask the transport to perform the resolution. In other words, we don’t need to add yet another component (pluggable resolvers): the transport can act as the resolver.

**Comment by Stebalien** [comment]
My concern is that resolvers and transports are independent. We'll want one resolver per name system (dns, dnsaddr, eth, handshake, etc.). This is is independent from the transports.

**Comment by raulk** [comment]
Understood. Under that model, resolvers would enlist themselves in a registry, and transports would call `multiaddr.Resolve()` which would use the global registry to traverse the multiaddr and resolve all resolvable components?

**Comment by Stebalien** [comment]
Resolvers would be registered with the swarm/host and the swarm/host would resolve all addresses, deduplicate the results, sort the addresses based on priority, then invoke the transports.

The transports can't be the ones to resolve the addresses because of all the reasons I've mentioned above (we may not know the transport till we resolve an address and transports can't deduplicate addresses or plan dials with multiple addresses).

**Comment by jacobheun** [comment]
The latest patch release of go-libp2p, https://github.com/libp2p/go-libp2p/releases/tag/v0.7.4, says this issue is fixed.
> This patch also fixes #841.

Can this be closed or are there remaining issues we need to track?

**Comment by Stebalien** [comment]
I've cut a release of quic-go that fixes this, so yes.
